xgcm/padding.py

```diff
@@ -0,0 +1,306 @@
+}
+
+
+def _maybe_swap_dimension_names(da, from_name, to_name):
+    # renames 1D slices and swaps dimension names for higher dimensional slices
+    if to_name in da.dims:
+        da = da.rename({to_name: to_name + "dummy"})
+        if from_name in da.dims:
+            da = da.rename({from_name: to_name})
+        da = da.rename({to_name + "dummy": from_name})
+    else:
+        da = da.rename({from_name: to_name})
+    return da
+
+
+def _pad_face_connections(da, grid, padding_width, **kwargs):
+
+    # Can I get rid of all the mess with the coordinates by stripping them here?
+    # The output of the padding will not fit any coordinates anymore anyways and
+    # after the ufunc is applied they will all be back on place. So lets make our
+    # lives easier here.
+    da = da.reset_coords(drop=True).reset_index(
+        [di for di in da.dims if di in da.coords], drop=True
+    )
+
+    # This method works really nicely if all the boundary widths have the same size.
+    # This is however not very common. We often have boundary_width with (0,1).
+    # I had a ton of trouble accomodating with convoluted logic. The new approach
+    # we find the largest boundary width value, and pad every boundary/axis with this max
+    # value. As a final step we trim the padded dataset according to the original boundary
+    # widths.
+
+    def _expand_boundary_width(padding_width):
+        all_widths = []
+        for widths in padding_width.values():
+            all_widths.extend(list(widths))
+        max_width = max(all_widths)
+        expanded_padding_width = {
+            k: (max_width, max_width) for k in padding_width.keys()
+        }
+        return expanded_padding_width
+
+    # !!! Hacky. This fixes the cubed-sphere test. I am not quite sure yet what the underlying cause is.
+    # TODO: Why is this problem not caught in the padding tests?
+    # I am concerned that we are really hardcoding the axes in here. Is there a way to make this more general?
+    padding_width = {axname: padding_width.get(axname, (0, 0)) for axname in ["X", "Y"]}
+    padding_width_original = {k: v for k, v in padding_width.items()}
+
+    padding_width = _expand_boundary_width(padding_width)
+
+    # The edges of the array which are not connected, need to be padded with the
+    # legacy padding. This also ensures that the resulting padded faces result in
+    # the same shape. We will pad everything now and then replace the connections.
+    # That might however not be the most computational efficient way to do it.
+
+    da_prepadded = _pad_basic(da, grid, padding_width, **kwargs)
+
+    pad_axes = padding_width.keys()
+
+    # i made some alterations to the grid init to store the
+    # following info on the grid level. We can discuss if that makes more sense
+    # TODO: Do we need that information on the axis still?
+    facedim = grid._facedim
+    connections = grid._connections
```

Comment:
If it can be stored in the `Axis` class it should, and then `Grid` just looks it up from the `Axis` objects it contains.